drivers/net/bnxt/bnxt_txr.c

```diff
@@ -281,34 +282,50 @@ static void bnxt_tx_cmp(struct bnxt_tx_queue *txq, int nr_pkts)
 static int bnxt_handle_tx_cp(struct bnxt_tx_queue *txq)
 {
 	struct bnxt_cp_ring_info *cpr = txq->cp_ring;
-	uint32_t raw_cons = cpr->cp_raw_cons;
-	uint32_t cons;
+	uint16_t cons = cpr->cp_cons;
+	bool v = cpr->v;
+	uint32_t last_cons = UINT32_MAX;
+	bool last_v = cpr->v;
 	int nb_tx_pkts = 0;
 	struct tx_cmpl *txcmp;
 
 	if ((txq->tx_ring->tx_ring_struct->ring_size -
 			(bnxt_tx_avail(txq->tx_ring))) >
 			txq->tx_free_thresh) {
 		while (1) {
-			cons = RING_CMP(cpr->cp_ring_struct, raw_cons);
+			NEXT_CMP(cpr, cons, v);
 			txcmp = (struct tx_cmpl *)&cpr->cp_desc_ring[cons];
 
-			if (!CMP_VALID(txcmp, raw_cons, cpr->cp_ring_struct))
+			if (!CMP_VALID(cons, v, cpr))
 				break;
 
 			if (CMP_TYPE(txcmp) == TX_CMPL_TYPE_TX_L2)
 				nb_tx_pkts++;
-			else
+			else {
 				RTE_LOG(DEBUG, PMD,
 						"Unhandled CMP type %02x\n",
 						CMP_TYPE(txcmp));
-			raw_cons = NEXT_RAW_CMP(raw_cons);
+
+				if (txcmp->flags_type & rte_cpu_to_le_16(1)) {
```

Comment:
Per hsi_struct_defs_dpdk.h:

        /*
         * This field indicates the exact type of the completion. By
         * convention, the LSB identifies the length of the record in
         * 16B units. Even values indicate 16B records. Odd values
         * indicate 32B records.
         */